File: test/system/continuous/agitator.pl
#! /usr/bin/env perl

# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


use POSIX qw(strftime);
use Cwd qw();

if(scalar(@ARGV) != 6 && scalar(@ARGV) != 4){
  print "Usage : agitator.pl <min sleep before kill in minutes>[:max sleep before kill in minutes] <min sleep before tup in minutes>[:max sleep before tup in minutes] hdfs_user accumulo_user [<min kill> <max kill>]\n";
  exit(1);
}

$myself=`whoami`;
chomp($myself);
$am_root=($myself eq 'root');

$cwd=Cwd::cwd();
$ACCUMULO_HOME=$cwd . '/../../..';
$HADOOP_PREFIX=$ENV{"HADOOP_PREFIX"};

print "Current directory: $cwd\n";
print "ACCUMULO_HOME=$ACCUMULO_HOME\n";
print "HADOOP_PREFIX=$HADOOP_PREFIX\n";

@sleeprange1 = split(/:/, $ARGV[0]);
$sleep1 = $sleeprange1[0];

@sleeprange2 = split(/:/, $ARGV[1]);
$sleep2 = $sleeprange2[0];

if (scalar(@sleeprange1) > 1) {
  $sleep1max = $sleeprange1[1] + 1;
} else {
  $sleep1max = $sleep1;
}

if ($sleep1 > $sleep1max) {
  die("sleep1 > sleep1max $sleep1 > $sleep1max");
}

if (scalar(@sleeprange2) > 1) {
  $sleep2max = $sleeprange2[1] + 1;
} else {
  $sleep2max = $sleep2;
}

if($sleep2 > $sleep2max){
  die("sleep2 > sleep2max $sleep2 > $sleep2max");
}

if(defined $ENV{'ACCUMULO_CONF_DIR'}){
  $ACCUMULO_CONF_DIR = $ENV{'ACCUMULO_CONF_DIR'};
}else{
  $ACCUMULO_CONF_DIR = $ACCUMULO_HOME . '/conf';
}

$HDFS_USER=$ARGV[2];
$ACCUMULO_USER=$ARGV[3];

$am_hdfs_user=($HDFS_USER eq $myself);
$am_accumulo_user=($ACCUMULO_USER eq $myself);

if(scalar(@ARGV) == 6){
  $minKill = $ARGV[4];
  $maxKill = $ARGV[5];
}else{
  $minKill = 1;
  $maxKill = 1;
}

if($minKill > $maxKill){
  die("minKill > maxKill $minKill > $maxKill");
}

@slavesRaw = `cat $ACCUMULO_CONF_DIR/slaves`;
chomp(@slavesRaw);

for $slave (@slavesRaw){
  if($slave eq "" || substr($slave,0,1) eq "#"){
    next;
  }

  push(@slaves, $slave);
}


if(scalar(@slaves) < $maxKill){
  print STDERR "WARN setting maxKill to ".scalar(@slaves)."\n";
  $maxKill = scalar(@slaves);
}

if ($minKill > $maxKill){
  print STDERR "WARN setting minKill to equal maxKill\n";
  $minKill = $maxKill;
}

while(1){

  $numToKill = int(rand($maxKill - $minKill + 1)) + $minKill;
  %killed = {};
  $server = "";
  $kill_tserver = 0;
  $kill_datanode = 0;

  for($i = 0; $i < $numToKill; $i++){
    while($server eq "" || $killed{$server} != undef){
      $index = int(rand(scalar(@slaves)));
      $server = $slaves[$index];
    }

    $killed{$server} = 1;

    $t = strftime "%Y%m%d %H:%M:%S", localtime;

    $rn = rand(1);
    if ($rn <.33) {
      $kill_tserver = 1;
      $kill_datanode = 1;
    } elsif ($rn < .66) {
      $kill_tserver = 1;
      $kill_datanode = 0;
    } else {
      $kill_tserver = 0;
      $kill_datanode = 1;
    }

    print STDERR "$t Killing $server $kill_tserver $kill_datanode\n";
    if ($kill_tserver) {
      if ($am_root) {
        # We're root, switch to the Accumulo user and try to stop gracefully
        system("su -c '$ACCUMULO_HOME/bin/stop-server.sh $server \"accumulo-start.jar\" tserver KILL' - $ACCUMULO_USER");
      } elsif ($am_accumulo_user) {
        # We're the accumulo user, just run the commandj
        system("$ACCUMULO_HOME/bin/stop-server.sh $server 'accumulo-start.jar' tserver KILL");
      } else {
        # We're not the accumulo user, try to use sudo
        system("sudo -u $ACCUMULO_USER $ACCUMULO_HOME/bin/stop-server.sh $server accumulo-start.jar tserver KILL");
      }
    }

    if ($kill_datanode) {
      if ($am_root) {
        # We're root, switch to HDFS to ssh and kill the process
        system("su -c 'ssh $server pkill -9 -f [p]roc_datanode' - $HDFS_USER");
      } elsif ($am_hdfs_user) {
        # We're the HDFS user, just kill the process
        system("ssh $server \"pkill -9 -f '[p]roc_datanode'\"");
      } else {
        # We're not the hdfs user, try to use sudo
        system("ssh $server 'sudo -u $HDFS_USER pkill -9 -f \'[p]roc_datanode\''");
      }
    }
  }

  $nextsleep2 = int(rand($sleep2max - $sleep2)) + $sleep2;
  sleep($nextsleep2 * 60);
  $t = strftime "%Y%m%d %H:%M:%S", localtime;
  print STDERR "$t Running tup\n";
  if ($am_root) {
    # Running as root, su to the accumulo user
    system("su -c $ACCUMULO_HOME/bin/tup.sh - $ACCUMULO_USER");
  } elsif ($am_accumulo_user) {
    # restart the as them as the accumulo user
    system("$ACCUMULO_HOME/bin/tup.sh");
  } else {
    # Not the accumulo user, try to sudo to the accumulo user
    system("sudo -u $ACCUMULO_USER $ACCUMULO_HOME/bin/tup.sh");
  }

  if ($kill_datanode) {
    print STDERR "$t Starting datanode on $server\n";
    if ($am_root) {
      # We're root, switch to the HDFS user
      system("ssh $server 'su -c \"$HADOOP_PREFIX/sbin/hadoop-daemon.sh start datanode\" - $HDFS_USER 2>/dev/null 1>/dev/null'");
    } elsif ($am_hdfs_user) {
      # We can just start as we're the HDFS user
      system("ssh $server '$HADOOP_PREFIX/sbin/hadoop-daemon.sh start datanode'");
    } else {
      # Not the HDFS user, have to try sudo
      system("ssh $server 'sudo -u $HDFS_USER $HADOOP_PREFIX/sbin/hadoop-daemon.sh start datanode'");
    }
  }

  $nextsleep1 = int(rand($sleep1max - $sleep1)) + $sleep1;
  sleep($nextsleep1 * 60);
}

File: modules/test/system/agitator/agitator.py
#! /usr/bin/python

# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import random
import logging
import ConfigParser

# add the environment variables as default settings
import os
defaults=dict([('env.' + k, v) for k, v in os.environ.iteritems()])
config = ConfigParser.ConfigParser(defaults)

# things you can do to a particular kind of process
class Proc:
   program = 'Unknown'
   _frequencyToKill = 1.0

   def start(self, host):
       pass

   def find(self, host):
       pass

   def numberToKill(self):
       return (1, 1)

   def frequencyToKill(self):
       return self._frequencyToKill

   def user(self):
       return config.get(self.program, 'user')

   def kill(self, host, pid):
      kill = config.get('agitator', 'kill').split()
      code, stdout, stderr = self.runOn(host, kill + [pid])
      if code != 0:
         raise logging.warn("Unable to kill %d on %s (%s)", pid, host, stderr)

   def runOn(self, host, cmd):
      ssh = config.get('agitator', 'ssh').split()
      return self.run(ssh + ["%s@%s" % (self.user(), host)] + cmd)

   def run(self, cmd):
      import subprocess
      cmd = map(str, cmd)
      logging.debug('Running %s', cmd)
      p = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
      stdout, stderr = p.communicate()
      if stdout.strip():
         logging.debug("%s", stdout.strip())
      if stderr.strip():
         logging.error("%s", stderr.strip())
      if p.returncode != 0:
         logging.error("Problem running %s", ' '.join(cmd))
      return p.returncode, stdout, stderr

   def __repr__(self):
      return self.program

class Zookeeper(Proc):
   program = 'zookeeper'
   def __init__(self):
      self._frequencyToKill = config.getfloat(self.program, 'frequency')

   def start(self, host):
      self.runOn(host, [config.get(self.program, 'home') + '/bin/zkServer.sh start'])

   def find(self, host):
     code, stdout, stderr = self.runOn(host, ['pgrep -f [Q]uorumPeerMain || true'])
     return map(int, [line for line in stdout.split("\n") if line])

class Hadoop(Proc):
   section = 'hadoop'
   def __init__(self, program):
      self.program = program
      self._frequencyToKill = config.getfloat(self.section, program + '.frequency')
      self.minimumToKill = config.getint(self.section, program + '.kill.min')
      self.maximumToKill = config.getint(self.section, program + '.kill.max')

   def start(self, host):
      binDir = config.get(self.section, 'bin')
      self.runOn(host, ['nohup %s/hdfs %s < /dev/null >/dev/null 2>&1 &' %(binDir, self.program)])
     
   def find(self, host):
      code, stdout, stderr = self.runOn(host, ["pgrep -f 'proc[_]%s' || true" % (self.program,)])
      return map(int, [line for line in stdout.split("\n") if line])

   def numberToKill(self):
      return (self.minimumToKill, self.maximumToKill)

   def user(self):
      return config.get(self.section, 'user')

class Accumulo(Hadoop):
   section = 'accumulo'
   def start(self, host):
      home = config.get(self.section, 'home')
      self.runOn(host, ['nohup %s/bin/accumulo %s </dev/null >/dev/null 2>&1 & ' %(home, self.program)])

   def find(self, host):
      code, stdout, stderr = self.runOn(host, ["pgrep -f 'app[=]%s' || true" % self.program])
      return map(int, [line for line in stdout.split("\n") if line])

def fail(msg):
   import sys
   logging.critical(msg)
   sys.exit(1)

def jitter(n):
   return random.random() * n - n / 2

def sleep(n):
   if n > 0:
       logging.info("Sleeping %.2f", n)
       import time
       time.sleep(n)

def agitate(hosts, procs):
   starters = []

   logging.info("Agitating %s on %d hosts" % (procs, len(hosts)))

   section = 'agitator'

   # repeatedly...
   while True:
      if starters:
         # start up services that were previously killed
         t = max(0, config.getfloat(section, 'sleep.restart') + jitter(config.getfloat(section, 'sleep.jitter')))
         sleep(t)
         for host, proc in starters:
            logging.info('Starting %s on %s', proc, host)
            proc.start(host)
         starters = []

      # wait some time
      t = max(0, config.getfloat(section, 'sleep') + jitter(config.getfloat(section, 'sleep.jitter')))
      sleep(t)

      # for some processes
      for p in procs:

         # roll dice: should it be killed?
         if random.random() < p.frequencyToKill():

            # find them
            from multiprocessing import Pool
            def finder(host):
               return host, p.find(host)
            with Pool(5) as pool:
               result = pool.map(finder, hosts)
            candidates = {}
            for host, pids in result:
               if pids:
                  candidates[host] = pids

            # how many?
            minKill, maxKill = p.numberToKill()
            count = min(random.randrange(minKill, maxKill + 1), len(candidates))

            # pick the victims
            doomedHosts = random.sample(candidates.keys(), count)

            # kill them
            logging.info("Killing %s on %s", p, doomedHosts)
            for doomedHost in doomedHosts:
               pids = candidates[doomedHost]
               if not pids:
                  logging.error("Unable to kill any %s on %s: no processes of that type are running", p, doomedHost)
               else:
                  pid = random.choice(pids)
                  logging.debug("Killing %s (%d) on %s", p, pid, doomedHost)
                  p.kill(doomedHost, pid)
                  # remember to restart them later
                  starters.append((doomedHost, p))

def main():
   import argparse
   parser = argparse.ArgumentParser(description='Kill random processes')
   parser.add_argument('--log', help='set the log level', default='INFO')
   parser.add_argument('--namenodes', help='randomly kill namenodes', action="store_true")
   parser.add_argument('--secondary', help='randomly kill secondary namenode', action="store_true")
   parser.add_argument('--datanodes', help='randomly kill datanodes', action="store_true")
   parser.add_argument('--tservers', help='randomly kill tservers', action="store_true")
   parser.add_argument('--masters', help='randomly kill masters', action="store_true")
   parser.add_argument('--zookeepers', help='randomly kill zookeepers', action="store_true")
   parser.add_argument('--gc', help='randomly kill the file garbage collector', action="store_true")
   parser.add_argument('--all', 
                       help='kill any of the tservers, masters, datanodes, namenodes or zookeepers', 
                       action='store_true')
   parser.add_argument('--hosts', type=argparse.FileType('r'), required=True)
   parser.add_argument('--config', type=argparse.FileType('r'), required=True)
   args = parser.parse_args()

   config.readfp(args.config)

   level = getattr(logging, args.log.upper(), None)
   if isinstance(level, int):
      logging.basicConfig(level=level)

   procs = []
   def addIf(flag, proc):
       if flag or args.all:
          procs.append(proc)

   addIf(args.namenodes,  Hadoop('namenode'))
   addIf(args.datanodes,  Hadoop('datanode'))
   addIf(args.secondary,  Hadoop('secondarynamenode'))
   addIf(args.tservers,   Accumulo('tserver'))
   addIf(args.masters,    Accumulo('master'))
   addIf(args.gc,         Accumulo('gc'))
   addIf(args.zookeepers, Zookeeper())
   if len(procs) == 0:
       fail("No processes to agitate!\n")

   hosts = []
   for line in args.hosts.readlines():
       line = line.strip()
       if line and line[0] != '#':
           hosts.append(line)
   if not hosts:
       fail('No hosts to agitate!\n')

   agitate(hosts, procs)

if __name__ == '__main__':
   main()
